tests/noviswitch.py

```diff
@@ -0,0 +1,604 @@
+import os
+import re
+import json
+import paramiko
+import time
+from mininet.node import Switch
+from mininet.clean import addCleanupCallback
+import logging
+import logging.handlers
+
+
+novi_cleanup_commands = [
+    "del config controller controllergroup all controllerid all",
+    "del config flow tableid all",
+]
+
+formatter = logging.Formatter(fmt="%(asctime)s %(name)s - %(levelname)s - %(message)s")
+handler = logging.handlers.SysLogHandler("/dev/log")
+handler.setFormatter(formatter)
+logger = logging.getLogger("noviswitch")
+logger.setLevel(logging.DEBUG)
+logger.addHandler(handler)
+
+
+def run_cmd(cmd):
+    logger.debug("running cmd in %s: %s" % (NOVILOCALIP, cmd))
+    stream = os.popen(cmd)
+    return stream.read()
+
+
+NOVISETTINGS = os.environ.get("NOVISETTINGS", {})
+NOVISWITCHES = os.environ.get("NOVISWITCHES")
+if NOVISETTINGS:
+    try:
+        NOVISETTINGS = json.load(open(NOVISETTINGS))
+    except:
+        try:
+            NOVISETTINGS = json.loads(NOVISETTINGS)
+        except:
+            raise ValueError(
+                "Invalid NOVISETTINGS environment variable: must be a JSON encoded string or file"
+            )
+elif NOVISWITCHES:
+    switches = NOVISWITCHES.split()
+    if len(switches) < 2:
+        raise ValueError(
+            "Invalid NOVISWITCHES environment variable: must be a space-separated list of IPs"
+        )
+    NOVISETTINGS = {}
+    for i, sw in enumerate(switches, 1):
+        NOVISETTINGS["vnovi%d" % i] = {"ip": sw}
+elif os.environ.get("SWITCH_CLASS") == "NoviSwitch":
+    raise ValueError(
+        "Neither NOVISWITCHES nor NOVISETTINGS environment variables found. You must define one of them"
+    )
+
+NOVIUSER = os.environ.get("NOVIUSER")
+NOVIPASS = os.environ.get("NOVIPASS")
+if os.environ.get("SWITCH_CLASS") == "NoviSwitch" and (not NOVIUSER or not NOVIPASS):
+    raise ValueError(
+        "Missing env vars for username/password for Noviflow switches: NOVIUSER and NOVIPASS"
+    )
+
+NOVILOCALIP = os.environ.get("NOVILOCALIP")
+if os.environ.get("SWITCH_CLASS") == "NoviSwitch" and not NOVILOCALIP:
+    sw = next(iter(NOVISETTINGS.values()))
+    NOVILOCALIP = run_cmd(f"ip -j route get {sw['ip']} | jq -r '.[0].prefsrc'")
+    if not NOVILOCALIP:
+        raise ValueError(
+            "Could not identify the Local IP for L2TP tunnels. Please define NOVILOCALIP env var or check pref ip source for switches"
+        )
+    NOVILOCALIP = NOVILOCALIP.strip()
+
+NOVIMAXWAIT = int(os.environ.get("NOVIMAXWAIT", 30))
+
+
+class NoviSwitch(Switch):
```

Comment:
Side comment: in the future, we should try to get this `NoviSwitch` e2e tested too, not necessarily even extracting as a full blown lib, but it'd be helpful to try to run basic e2e test for it too to know that the ssh wrapper is behaving as you'd expect (this class reponsibility/complexity is already significant), so when things go wrong it'd be easier for us to figure it out. This could be a future project for a capstone student, if getting them access to the virtual noviflow switch isn't too difficult in a AmLight server. Wdyt, should we map this in the backlog?